src/components/paster.ts

```diff
@@ -0,0 +1,201 @@
+                const rows: string[] = []
+
+                fs.createReadStream(path.resolve(baseFile, file))
+                    .pipe(csv())
+                    .on('data', (data: Object) => rows.push(Object.values(data).join('\t')))
+                    .on('end', () => {
+                        const body = rows.join('\n')
+                        this.pasteTable(editor, body)
+                    })
+            }
+        }
+    }
+
+    public pasteTable(editor: vscode.TextEditor, content: string) {
+        this.extension.logger.addLogMessage('Pasting: Table')
+        // trim surrounding whitespace
+        content = content.replace(/^\s*/, '').replace(/\s*$/, '')
+        content = this.reformatText(content, false)
+        const lines = content.split('\n')
+        const cells = lines.map(l => l.split('\t'))
+
+        // determine if all rows have same number of cells
+        const isConsistent = cells.reduce((accumulator, current, _index, array) => {
+            if (current.length === array[0].length) {
+                return accumulator
+            } else {
+                return false
+            }
+        }, true)
+        if (!isConsistent) {
+            throw new Error('Table is not consistent')
+        } else if (cells.length === 1 && cells[0].length === 1) {
+            this.pasteNormal(editor, content)
+            return
+        }
+
+        const columnType: string = vscode.workspace.getConfiguration('latex-workshop.formattedPaste')['tableColumnType']
+        const booktabs: boolean = vscode.workspace.getConfiguration('latex-workshop.formattedPaste')['tableBooktabsStyle']
+        const headerRows: number = vscode.workspace.getConfiguration('latex-workshop.formattedPaste')['tableHeaderRows']
+        const tabularRows = cells.map(row => '\t' + row.join(' & '))
+
+        if (headerRows && tabularRows.length > headerRows) {
+            const eol = editor.document.eol === vscode.EndOfLine.LF ? '\n' : '\r\n'
+            const headSep = '\t' + (booktabs ? '\\midrule' : '\\hline') + eol
+            tabularRows[headerRows] = headSep + tabularRows[headerRows]
+        }
+        let tabularContents = tabularRows.join(' \\\\\n')
+        if (booktabs) {
+            tabularContents = '\t\\toprule\n' + tabularContents + ' \\\\\n\t\\bottomrule'
+        }
+        const tabular = `\\begin{tabular}{${columnType.repeat(cells[0].length)}}\n${tabularContents}\n\\end{tabular}`
+
+        editor.edit(edit => {
+            const current = editor.selection
+
+            if (current.isEmpty) {
+                edit.insert(current.start, tabular)
+            } else {
+                edit.replace(current, tabular)
+            }
+        })
+    }
+
+    public reformatText(text: string, removeBonusWhitespace = true) {
+        function doRemoveBonusWhitespace(str: string) {
+            str = str.replace(/\n\n/g, '\uE000')
+            str = str.replace(/\s+/g, ' ')
+            str = str.replace(/\uE000/g, '\n\n')
```

Comment:
What is the point of `str = str.replace(/\n\n/g, '\uE000')` and then `str = str.replace(/\uE000/g, '\n\n')`?

Reply:
Something I've noticed a fair bit when pasting from various sources is a lot of bogus newlines and spaces appear. White this doesn't affect the output it would be nicer to not have this happen. Sometimes pastes have special stuff such as weird spaces which can be a bit annoying. Hence I want to replace all white space with plain ol' breaking, singlewidth spaces.

The one exception to this is paragraphs, which are separated by a double newline. Hence I replace double new lines with a private use character (I think it's reasonable to expect this to not appear in any strings), and then turn those back into double newlines.

Thinking a bit more about this a few minor improvements could be made though. See a548e342bc4e6b6770f7aa42a5c39c84a6519338.